### DISCUSSION TITLE
Is there a way to bind insert mode key-press to a command?

### DISCUSSION THREAD

**Discussion by rkshthrmsh** [helix-editor/helix#8747] [Q&A] [answered]
I am attempting to remap `[<space>` in normal mode to insert space before the cursor / selection. I also want to remap `add_newline_above` to `[n`. The second part is straightforward enough, but how can the first one be done? Here's my attempt:

```
[keys.normal."["]
space = ["insert_mode", <???>, "normal_mode"]
n = "add_newline_above"
```

**Comment by the-mikedavis** [comment]
This would be covered by macro keybinding in the future: #1383. With that feature implemented, the middle command would be `"@<space>"` (`@` meaning that the command is a macro and `<space>` to insert the space).

**Reply by Omnikar** [reply]
I was going to mention the workaround using `:insert-output echo …`, but it seems that due to a limitation in the way Helix parses shell commands, it's not really possible to pass a space as an argument, i.e. `:insert-output echo -n ' '` doesn't work.

**Reply by the-mikedavis** [reply]
Looks like shellwords are eating the single quotes, you might be able to just escape them: `:insert-output echo -n \' \'`

**Reply by Omnikar** [reply]
Oh, that worked. But I also tried it as a keybind config and it seems it actually works as you'd expect for that. I'll make a new answer suggestion with that then.

**Comment by Omnikar** [comment]
Definitely look at what @the-mikedavis said, but here's a workaround:
```
space = ":insert-output echo -n ' '"
```

**Comment by rkshthrmsh** [comment]
@Omnikar and @the-mikedavis, (a bit late, but) it looks like `:insert-output echo -n ' '` is inserting `-n` and a space. This is happening even from the command mode within helix.

**Reply by Omnikar** [reply, marked as answer]
Try `:insert-output printf ' '` instead
